tools/flowc/statements/fs_inline.flow

```diff
@@ -0,0 +1,599 @@
+FsInlineOperator(
+	fnArg: ([FsExp]) -> FsVarUse,
+	inline: (FsStat, [FsExp], FsClosure, () -> string) -> FsStat
+);
+
+fsInlineStatsDo(s: FsStat,
+	worth_inlining: (FsAll) -> bool,
+	closure_lets: Tree<string, FsLet>,
+	inlined_names: ref Set<string>,
+	new_id: () -> string, debug: bool
+) -> FsStat {
+	try_exp = \x: FsStat, e: FsExp -> switch(e) {
+		FsCall(o, args, __,__): {
+			switch (lookupTree(fs_inlined_operators, o.var.name)) {
+				Some(op): {
+					fn = op.fnArg(args);
+					switch (lookupTree(closure_lets, fn.var.name)) {
+						Some(let): {
+							if (!fsIsReacheableByLetSeq(let, x.id)) {
+								if (debug) {
+									println(i2s(x.id) + " is not reacheable from " + i2s(let.id));
+								}
+								x
+							} else {
+								inlined_names := insertSet(^inlined_names, let.var.name);
+								op.inline(x, args, cast(let.e: FsExp -> FsClosure), new_id)
+							}
+						}
+						None(): x;
+					}
+				}
+				None(): x;
+			}
+		}
+		default: x;
+	}
+	s2 = switch (s) {
+		FsLet(v, e, s0, id,__): {
+			e1 = switch (e) {
+				FsClosure(vars, lambda,__,__): {
+					FsClosure(e with
+						lambda = FsLambda(lambda with
+							body = fsInlineStatsDo(lambda.body, worth_inlining, makeTree(), inlined_names, new_id, debug)
+						)
+					);
+				}
+				default: e;
+			}
+			s1 = FsLet(s with
+				e = e1,
+				s = fsInlineStatsDo(s0, worth_inlining,
+					switch (e) {
+						FsClosure(__,lambda,__,__): {
+							if (!worth_inlining(lambda.body)) {
+								if (debug) {
+									println("not worth ninining, id: " + i2s(id));
+								}
+								closure_lets
+							} else {
+								setTree(closure_lets, v.name, s);
+							}
+						}
+						default: closure_lets;
+					},
+					inlined_names, new_id, debug
+				)
+			);
+			try_exp(s1, e1);
+		}
+		FsIf(cond, s1, s2,__,__): {
+			FsIf(s with
+				s1 = fsInlineStatsDo(s1, worth_inlining, closure_lets, inlined_names, new_id, debug),
+				s2 = fsInlineStatsDo(s2, worth_inlining, closure_lets, inlined_names, new_id, debug),
+			);
+		}
+		FsSwitch(__,__,cs,__,nx): {
+			FsSwitch(s with
+				cases = map(cs, \c ->
+					FsCase(c with body = fsInlineStatsDo(c.body, worth_inlining, closure_lets, inlined_names, new_id, debug))
+				),
+			);
+		}
+		FsFor(v, arr, s1,__,__,nx): {
+			FsFor(s with
+				s = fsInlineStatsDo(s1, worth_inlining, closure_lets, inlined_names, new_id, debug),
+			);
+		}
+		FsRet(e,__,__): {
+			switch (e) {
+				FsClosure(__,lambda,__,__): {
+					e1 = FsClosure(e with
+						lambda = FsLambda(lambda with
+							body = fsInlineStatsDo(lambda.body, worth_inlining, makeTree(), inlined_names, new_id, debug)
+						)
+					);
+					FsRet(s with e = e1);
+				}
+				default: try_exp(s, e);
+			}
+		}
+		FsWrapExp(e,__,__): try_exp(s, e);
+		default: s;
+	}
+	switch (s2.next) {
+		Some(nx): fsSetNext(s2, fsInlineStatsDo(nx, worth_inlining, closure_lets, inlined_names, new_id, debug));
+		None(): s2;
+	}
+}
+
+fs_inlined_operators = pairs2tree([
+	Pair("fold", FsInlineOperator(
```

Comment:
Can we define some listf of names in flow.config to define what functions to inline? It seems we can automatically detect the lambda args from the types?